eth_tx_trie.go

```diff
@@ -0,0 +1,277 @@
+// It is used by both the trie calculations on block body parsing,
+// as well as by the decoding of an object from the IPLD forest.
+func DecodeEthTxTrie(c *cid.Cid, b []byte) (*EthTxTrie, error) {
+	var i []interface{}
+	err := rlp.DecodeBytes(b, &i)
+	if err != nil {
+		return nil, err
+	}
+
+	switch len(i) {
+	case 2:
+		key := i[0].([]byte)
+		valb := i[1].([]byte)
+
+		var val interface{}
+		if len(valb) == 32 {
+			// This is a reference to another eth-tx-trie object
+			val = keccak256ToCid(MEthTxTrie, valb)
+		} else {
+			// This is a proper transaction
+			var t types.Transaction
+			err := rlp.DecodeBytes(valb, &t)
+			if err != nil {
+				return nil, err
+			}
+			val = &EthTx{
+				Transaction: &t,
+				cid:         rawdataToCid(MEthTx, valb),
+				rawdata:     valb,
+			}
+		}
+		return &EthTxTrie{
+			arr:     []interface{}{key, val},
+			cid:     c,
+			rawdata: b,
+		}, nil
+	case 17:
+		var parsed []interface{}
+		for _, v := range i {
+			bv := v.([]byte)
+			switch len(bv) {
+			case 0:
+				parsed = append(parsed, nil)
+			case 32:
+				parsed = append(parsed, keccak256ToCid(MEthTxTrie, bv))
+			default:
+				return nil, fmt.Errorf("unrecognized object in trie: %v", bv)
+			}
+		}
+		return &EthTxTrie{
+			arr:     parsed,
+			rawdata: b,
+			cid:     c,
+		}, nil
+	default:
+		return nil, fmt.Errorf("unknown trie node type")
+	}
+}
+
+/*
+  Block INTERFACE
+*/
+
+// RawData returns the binary of the RLP encode of the transaction.
+func (t *EthTxTrie) RawData() []byte {
+	return t.rawdata
+}
+
+// Cid returns the cid of the transaction.
+func (t *EthTxTrie) Cid() *cid.Cid {
+	return t.cid
+}
+
+// String is a helper for output
+func (t *EthTxTrie) String() string {
+	return fmt.Sprintf("<EthereumTxTrie %s>", t.cid)
+}
+
+// Loggable returns in a map the type of IPLD Link.
+func (t *EthTxTrie) Loggable() map[string]interface{} {
+	return map[string]interface{}{
+		"type": "eth-tx-trie",
+	}
+}
+
+// Resolve resolves a path through this node, stopping at any link boundary
+// and returning the object found as well as the remaining path to traverse
+func (t *EthTxTrie) Resolve(p []string) (interface{}, []string, error) {
+	if len(p) == 0 {
+		return t, nil, nil
+	}
+
+	i, err := strconv.Atoi(p[0])
+	if err != nil {
+		return nil, nil, fmt.Errorf("expected array index to trie: %s", err)
+	}
+
+	if i < 0 || i >= len(t.arr) {
+		return nil, nil, fmt.Errorf("index in trie out of range")
+	}
+
+	switch obj := t.arr[i].(type) {
+	case *cid.Cid:
+		return &node.Link{Cid: obj}, p[1:], nil
+	case *EthTx:
+		return obj, p[1:], nil
+	default:
+		return nil, nil, fmt.Errorf("unexpected object type in trie")
+	}
+}
+
+// Tree lists all paths within the object under 'path', and up to the given depth.
+// To list the entire object (similar to `find .`) pass "" and -1
+func (t *EthTxTrie) Tree(p string, depth int) []string {
+	if p != "" {
+		return nil
+	}
+	if depth > 0 {
```

Comment:
Oh. That was "inherited" (https://github.com/ipfs/go-ipld-eth/blob/d683c34dbd034547d9c1b911b9341fe82bd4ab64/trie.go). I haven't completed my go over the eth-tx-trie elements yet.